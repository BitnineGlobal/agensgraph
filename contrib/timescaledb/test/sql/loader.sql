\c single :ROLE_SUPERUSER

DROP EXTENSION timescaledb;
--no extension
\dx
SELECT 1;

CREATE EXTENSION timescaledb VERSION 'mock-1';
SELECT 1;
\dx

CREATE EXTENSION IF NOT EXISTS timescaledb VERSION 'mock-1';
CREATE EXTENSION IF NOT EXISTS timescaledb VERSION 'mock-2';

DROP EXTENSION timescaledb;
\set ON_ERROR_STOP 0
--test that we cannot accidentally load another library version
CREATE EXTENSION IF NOT EXISTS timescaledb VERSION 'mock-2';
\set ON_ERROR_STOP 1

\c single :ROLE_SUPERUSER
--no extension
\dx
SELECT 1;

CREATE EXTENSION timescaledb VERSION 'mock-1';
--same backend as create extension;
SELECT 1;
\dx

--start new backend;
\c single :ROLE_DEFAULT_PERM_USER

SELECT 1;
SELECT 1;
--test fn call after load
SELECT mock_function();
\dx

\c single :ROLE_DEFAULT_PERM_USER
--test fn call as first command
SELECT mock_function();

--use guc to prevent loading
\c single :ROLE_SUPERUSER
SET timescaledb.disable_load = 'on';
SELECT 1;
SELECT 1;
SET timescaledb.disable_load = 'off';
SELECT 1;
\set ON_ERROR_STOP 0
SET timescaledb.disable_load = 'not bool';
\set ON_ERROR_STOP 1

\c single :ROLE_SUPERUSER
RESET ALL;
SELECT 1;

\c single :ROLE_SUPERUSER
SET timescaledb.disable_load TO DEFAULT;
SELECT 1;

\c single :ROLE_SUPERUSER
RESET timescaledb.disable_load;
SELECT 1;

\c single :ROLE_SUPERUSER
SET timescaledb.other = 'on';
SELECT 1;

\set ON_ERROR_STOP 0
--cannot update extension after .so of previous version already loaded
ALTER EXTENSION timescaledb UPDATE TO 'mock-2';
\set ON_ERROR_STOP 1

\c single_2 :ROLE_SUPERUSER
\dx
CREATE EXTENSION timescaledb VERSION 'mock-1';
\dx
--start a new backend to update
\c single_2 :ROLE_SUPERUSER
ALTER EXTENSION timescaledb UPDATE TO 'mock-2';
SELECT 1;
\dx

--drop extension
DROP EXTENSION timescaledb;
SELECT 1;
\dx


\c single_2 :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-2';
SELECT 1;
\dx

--single still has old version
\c single :ROLE_SUPERUSER
SELECT 1;
\dx

--try a broken upgrade
\c single_2 :ROLE_SUPERUSER
\dx
\set ON_ERROR_STOP 0
ALTER EXTENSION timescaledb UPDATE TO 'mock-3';
\set ON_ERROR_STOP 1
--should still be on mock-2
SELECT 1;
\dx

--drop extension
DROP EXTENSION timescaledb;
SELECT 1;
\dx

--create extension anew, only upgrade was broken
\c single_2 :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-3';
SELECT 1;
\dx
DROP EXTENSION timescaledb;
SELECT 1;

--mismatched version errors
\c single_2 :ROLE_SUPERUSER
\set ON_ERROR_STOP 0
--mock-4 has mismatched versions, so the .so load should throw an error
CREATE EXTENSION timescaledb VERSION 'mock-4';
\set ON_ERROR_STOP 1
--mock-4 not installed.
\dx

\set ON_ERROR_STOP 0
--should not allow since the errored-out mock-4 above already poisoned the well.
CREATE EXTENSION timescaledb VERSION 'mock-5';
\set ON_ERROR_STOP 1

\c single_2 :ROLE_SUPERUSER
--broken version and drop
CREATE EXTENSION timescaledb VERSION 'mock-broken';

\set ON_ERROR_STOP 0
--intentional broken version
\dx
SELECT 1;
SELECT 1;
--cannot drop extension; already loaded broken version
DROP EXTENSION timescaledb;
\set ON_ERROR_STOP 1

\c single_2 :ROLE_SUPERUSER
--can drop extension now. Since drop first command.
DROP EXTENSION timescaledb;
\dx

--broken version and update to fixed
\c single_2 :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-broken';
\set ON_ERROR_STOP 0
--intentional broken version
SELECT 1;
--cannot update extension; already loaded bad version
ALTER EXTENSION timescaledb UPDATE TO 'mock-5';
\set ON_ERROR_STOP 1

\c single_2 :ROLE_SUPERUSER
--can update extension now.
ALTER EXTENSION timescaledb UPDATE TO 'mock-5';
SELECT 1;
SELECT mock_function();

\c single_2 :ROLE_SUPERUSER
ALTER EXTENSION timescaledb UPDATE TO 'mock-6';
\set ON_ERROR_STOP 0
--The mock-5->mock_6 upgrade is intentionally broken.
--The mock_function was never changed to point to mock-6 in the update script.
--Thus mock_function is defined incorrectly to point to the mock-5.so
--This should be an error.
SELECT mock_function();
\set ON_ERROR_STOP 1
\dx

--TEST: create extension when old .so already loaded
\c single :ROLE_SUPERUSER
--force load of extension with (\dx)
\dx
DROP EXTENSION timescaledb;
\dx

\set ON_ERROR_STOP 0
CREATE EXTENSION timescaledb VERSION 'mock-2';
\set ON_ERROR_STOP 1
\dx
--can create in a new session.
\c single :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-2';
\dx

--make sure parallel workers started after a 'DISCARD ALL' work
CREATE TABLE test (i int, j double precision);
INSERT INTO test SELECT x, x+0.1 FROM generate_series(1,100) AS x;

DISCARD ALL;
SET force_parallel_mode = 'on';
SET max_parallel_workers_per_gather = 1;
SELECT count(*) FROM test;
