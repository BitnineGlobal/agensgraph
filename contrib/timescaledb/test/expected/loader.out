\c single :ROLE_SUPERUSER
DROP EXTENSION timescaledb;
--no extension
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

SELECT 1;
 ?column? 
----------
        1
(1 row)

CREATE EXTENSION timescaledb VERSION 'mock-1';
WARNING:  mock init "mock-1"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-1"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-1  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

CREATE EXTENSION IF NOT EXISTS timescaledb VERSION 'mock-1';
WARNING:  mock post_analyze_hook "mock-1"
NOTICE:  extension "timescaledb" already exists, skipping
CREATE EXTENSION IF NOT EXISTS timescaledb VERSION 'mock-2';
WARNING:  mock post_analyze_hook "mock-1"
NOTICE:  extension "timescaledb" already exists, skipping
DROP EXTENSION timescaledb;
WARNING:  mock post_analyze_hook "mock-1"
\set ON_ERROR_STOP 0
--test that we cannot accidentally load another library version
CREATE EXTENSION IF NOT EXISTS timescaledb VERSION 'mock-2';
ERROR:  extension "timescaledb" has already been loaded with another version
\set ON_ERROR_STOP 1
\c single :ROLE_SUPERUSER
--no extension
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

SELECT 1;
 ?column? 
----------
        1
(1 row)

CREATE EXTENSION timescaledb VERSION 'mock-1';
WARNING:  mock init "mock-1"
--same backend as create extension;
SELECT 1;
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-1"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-1  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--start new backend;
\c single :ROLE_DEFAULT_PERM_USER
SELECT 1;
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

SELECT 1;
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

--test fn call after load
SELECT mock_function();
WARNING:  mock post_analyze_hook "mock-1"
WARNING:  mock function call "mock-1"
 mock_function 
---------------
 
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-1"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-1  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

\c single :ROLE_DEFAULT_PERM_USER
--test fn call as first command
SELECT mock_function();
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
WARNING:  mock function call "mock-1"
 mock_function 
---------------
 
(1 row)

--use guc to prevent loading
\c single :ROLE_SUPERUSER
SET timescaledb.disable_load = 'on';
SELECT 1;
 ?column? 
----------
        1
(1 row)

SELECT 1;
 ?column? 
----------
        1
(1 row)

SET timescaledb.disable_load = 'off';
SELECT 1;
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\set ON_ERROR_STOP 0
SET timescaledb.disable_load = 'not bool';
WARNING:  mock post_analyze_hook "mock-1"
ERROR:  parameter "timescaledb.disable_load" requires a Boolean value
\set ON_ERROR_STOP 1
\c single :ROLE_SUPERUSER
RESET ALL;
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\c single :ROLE_SUPERUSER
SET timescaledb.disable_load TO DEFAULT;
SELECT 1;
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\c single :ROLE_SUPERUSER
RESET timescaledb.disable_load;
SELECT 1;
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\c single :ROLE_SUPERUSER
SET timescaledb.other = 'on';
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\set ON_ERROR_STOP 0
--cannot update extension after .so of previous version already loaded
ALTER EXTENSION timescaledb UPDATE TO 'mock-2';
ERROR:  extension "timescaledb" cannot be updated after the old version has already been loaded
\set ON_ERROR_STOP 1
\c single_2 :ROLE_SUPERUSER
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

CREATE EXTENSION timescaledb VERSION 'mock-1';
WARNING:  mock init "mock-1"
\dx
WARNING:  mock post_analyze_hook "mock-1"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-1  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--start a new backend to update
\c single_2 :ROLE_SUPERUSER
ALTER EXTENSION timescaledb UPDATE TO 'mock-2';
WARNING:  mock init "mock-2"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-2"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-2"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-2  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--drop extension
DROP EXTENSION timescaledb;
WARNING:  mock post_analyze_hook "mock-2"
SELECT 1;
 ?column? 
----------
        1
(1 row)

\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

\c single_2 :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-2';
WARNING:  mock init "mock-2"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-2"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-2"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-2  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--single still has old version
\c single :ROLE_SUPERUSER
SELECT 1;
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-1"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-1  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--try a broken upgrade
\c single_2 :ROLE_SUPERUSER
\dx
WARNING:  mock init "mock-2"
WARNING:  mock post_analyze_hook "mock-2"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-2  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

\set ON_ERROR_STOP 0
ALTER EXTENSION timescaledb UPDATE TO 'mock-3';
ERROR:  extension "timescaledb" cannot be updated after the old version has already been loaded
\set ON_ERROR_STOP 1
--should still be on mock-2
SELECT 1;
WARNING:  mock post_analyze_hook "mock-2"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-2"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-2  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--drop extension
DROP EXTENSION timescaledb;
WARNING:  mock post_analyze_hook "mock-2"
SELECT 1;
 ?column? 
----------
        1
(1 row)

\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

--create extension anew, only upgrade was broken
\c single_2 :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-3';
WARNING:  mock init "mock-3"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-3"
 ?column? 
----------
        1
(1 row)

\dx
WARNING:  mock post_analyze_hook "mock-3"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-3  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

DROP EXTENSION timescaledb;
WARNING:  mock post_analyze_hook "mock-3"
SELECT 1;
 ?column? 
----------
        1
(1 row)

--mismatched version errors
\c single_2 :ROLE_SUPERUSER
\set ON_ERROR_STOP 0
--mock-4 has mismatched versions, so the .so load should throw an error
CREATE EXTENSION timescaledb VERSION 'mock-4';
ERROR:  extension "timescaledb" version mismatch: shared library version mock-4-mismatch; SQL version mock-4
\set ON_ERROR_STOP 1
--mock-4 not installed.
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

\set ON_ERROR_STOP 0
--should not allow since the errored-out mock-4 above already poisoned the well.
CREATE EXTENSION timescaledb VERSION 'mock-5';
ERROR:  extension "timescaledb" has already been loaded with another version
\set ON_ERROR_STOP 1
\c single_2 :ROLE_SUPERUSER
--broken version and drop
CREATE EXTENSION timescaledb VERSION 'mock-broken';
WARNING:  mock init "mock-broken"
\set ON_ERROR_STOP 0
--intentional broken version
\dx
WARNING:  mock post_analyze_hook "mock-broken"
ERROR:  mock broken "mock-broken"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-broken"
ERROR:  mock broken "mock-broken"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-broken"
ERROR:  mock broken "mock-broken"
--cannot drop extension; already loaded broken version
DROP EXTENSION timescaledb;
WARNING:  mock post_analyze_hook "mock-broken"
ERROR:  mock broken "mock-broken"
\set ON_ERROR_STOP 1
\c single_2 :ROLE_SUPERUSER
--can drop extension now. Since drop first command.
DROP EXTENSION timescaledb;
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

--broken version and update to fixed
\c single_2 :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-broken';
WARNING:  mock init "mock-broken"
\set ON_ERROR_STOP 0
--intentional broken version
SELECT 1;
WARNING:  mock post_analyze_hook "mock-broken"
ERROR:  mock broken "mock-broken"
--cannot update extension; already loaded bad version
ALTER EXTENSION timescaledb UPDATE TO 'mock-5';
ERROR:  extension "timescaledb" cannot be updated after the old version has already been loaded
\set ON_ERROR_STOP 1
\c single_2 :ROLE_SUPERUSER
--can update extension now.
ALTER EXTENSION timescaledb UPDATE TO 'mock-5';
WARNING:  mock init "mock-5"
SELECT 1;
WARNING:  mock post_analyze_hook "mock-5"
 ?column? 
----------
        1
(1 row)

SELECT mock_function();
WARNING:  mock post_analyze_hook "mock-5"
WARNING:  mock function call "mock-5"
 mock_function 
---------------
 
(1 row)

\c single_2 :ROLE_SUPERUSER
ALTER EXTENSION timescaledb UPDATE TO 'mock-6';
WARNING:  mock init "mock-6"
\set ON_ERROR_STOP 0
--The mock-5->mock_6 upgrade is intentionally broken.
--The mock_function was never changed to point to mock-6 in the update script.
--Thus mock_function is defined incorrectly to point to the mock-5.so
--This should be an error.
SELECT mock_function();
WARNING:  mock post_analyze_hook "mock-6"
ERROR:  extension "timescaledb" version mismatch: shared library version mock-5; SQL version mock-6
\set ON_ERROR_STOP 1
\dx
WARNING:  mock post_analyze_hook "mock-6"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-6  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--TEST: create extension when old .so already loaded
\c single :ROLE_SUPERUSER
--force load of extension with (\dx)
\dx
WARNING:  mock init "mock-1"
WARNING:  mock post_analyze_hook "mock-1"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-1  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

DROP EXTENSION timescaledb;
WARNING:  mock post_analyze_hook "mock-1"
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

\set ON_ERROR_STOP 0
CREATE EXTENSION timescaledb VERSION 'mock-2';
ERROR:  extension "timescaledb" has already been loaded with another version
\set ON_ERROR_STOP 1
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(1 row)

--can create in a new session.
\c single :ROLE_SUPERUSER
CREATE EXTENSION timescaledb VERSION 'mock-2';
WARNING:  mock init "mock-2"
\dx
WARNING:  mock post_analyze_hook "mock-2"
                                      List of installed extensions
    Name     | Version |   Schema   |                            Description                            
-------------+---------+------------+-------------------------------------------------------------------
 plpgsql     | 1.0     | pg_catalog | PL/pgSQL procedural language
 timescaledb | mock-2  | public     | Enables scalable inserts and complex queries for time-series data
(2 rows)

--make sure parallel workers started after a 'DISCARD ALL' work
CREATE TABLE test (i int, j double precision);
WARNING:  mock post_analyze_hook "mock-2"
INSERT INTO test SELECT x, x+0.1 FROM generate_series(1,100) AS x;
WARNING:  mock post_analyze_hook "mock-2"
DISCARD ALL;
WARNING:  mock post_analyze_hook "mock-2"
SET force_parallel_mode = 'on';
WARNING:  mock post_analyze_hook "mock-2"
SET max_parallel_workers_per_gather = 1;
WARNING:  mock post_analyze_hook "mock-2"
SELECT count(*) FROM test;
WARNING:  mock post_analyze_hook "mock-2"
WARNING:  mock init "mock-2"
 count 
-------
   100
(1 row)

